torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -101,6 +103,178 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
   }
 }
 
+namespace {
+
+// Helper for wireDeserialize() below.
+//
+// The format we use below looks like:
+//    section_name_1 size_1\n
+//    section_name_2 size_2\n
+//    ..
+//    \n
+//    [sections in order]
+//
+// Sections themselves include:
+//    - "payload" - the payload bits
+//    - "meta"    - metadata for the unpickler
+//    - "0" ...   - tensor sections for the unpickler
+//
+// Note that per the header comments, the format is subject to change,
+// and is best used for rpcs, rather than persistent disk storage.
+std::unordered_map<std::string, std::pair<const char*, size_t>>
+parseWireSections(const void* data, size_t data_size) {
+  const char* ptr = static_cast<const char*>(data);
+  const char* endp = ptr + data_size;
+
+  std::vector<std::pair<std::string, size_t>> header_ents;
+  bool ok = false;
+  while (ptr < endp) {
+    if (*ptr == '\n') {
+      ok = true;
+      break;
+    }
+    const char* num_ptr = ptr;
+    while (*ptr != ' ' && ptr < endp) {
+      ptr++;
+    }
+    if (ptr >= endp) {
+      break;
+    }
+    std::string name(num_ptr, ptr - num_ptr);
+    if (++ptr >= endp) {
+      break; // past the ' '
+    }
+    const char* size_ptr = ptr;
+    while (*ptr != '\n' && ptr < endp) {
+      ptr++;
+    }
+    if (ptr >= endp) {
+      break;
+    }
+    size_t sz = c10::stoll(std::string(size_ptr, ptr - size_ptr));
+    header_ents.push_back({name, sz});
+    ++ptr; // past the '\n'
+  }
+  ++ptr; // past the final \n
+
+  std::unordered_map<std::string, std::pair<const char*, size_t>> out;
+  for (size_t i = 0; i < header_ents.size(); ++i) {
+    out[header_ents[i].first] = {ptr, header_ents[i].second};
+    ptr += header_ents[i].second;
+  }
+  if (!ok || ptr > endp) {
+    throw std::runtime_error("failed bounds");
+  }
+  return out;
+}
+
+static const char* kMeta = "meta";
+static const char* kPayload = "payload";
+}; // namespace
+
+std::string wireSerialize(
+    const std::vector<char>& payload,
+    const std::vector<at::Tensor>& tensors) {
+  struct Ent {
+    std::string name;
+    const char* data;
+    size_t size;
+  };
+  std::vector<Ent> entries;
+  std::string metaEntry;
+  if (!payload.empty()) {
+    entries.push_back({kPayload, payload.data(), payload.size()});
+  }
+
+  if (!tensors.empty()) {
+    torch::jit::Pickler pickler(
+        [&](const void* buf, size_t sz) -> size_t {
+          metaEntry.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
+    pickler.protocol();
+    pickler.pushIValue(tensors);
```

Comment:
Question: how does `wireSerialize` compare to directly push everything into the jit pickler?

Reply:
My understanding here is that the main viable alternative, if you need to send over the wire, is effectively torch::save(), which is several times slower, per the diff description.

If you provide a tensorTable, it will just push_back into the table (via pushTensorReference()), but not with metadata or anything useful - you're still stuck where you began with respect to getting these tensors in a blittable format.

Here, we're basically using the pickler with the tensors like ScriptModuleSerializer::writeArchive() does (see torch/csrc/jit/export.cpp) as part of torch::save(), so it's a fairly well-used codepath. The ScriptModuleSerializer use of this codepath then just puts the tensor files "0", "1", "2", etc in a zip archive, while we're doing something quite similar, but in a more streamlined manner.

The efficiency gains come chiefly from
  - avoiding CRC overhead (~35% of time, even with non-open-sourced 40x faster CRC function)
  - avoiding unnecessary copies and L2 miss that come from zip format
  - avoiding sending a per-call python program that takes ~15usec to Lex/Parse in every torch::load() call, even for a minimal tensor